apps/cowswap-frontend/src/modules/twap/utils/getTwapOrderStatus.ts

```diff
@@ -37,3 +39,53 @@ export function isTwapOrderExpired(order: TWAPOrderStruct, startDate: Date | nul
 
   return nowTimestamp > endTime
 }
+
+export function getTwapParentStatusFromChildren(
+  childrenOrders: { status: OrderStatus; isCancelling?: boolean }[],
+): TwapOrderStatus {
+  if (!childrenOrders.length) return TwapOrderStatus.Pending
+
+  const hasOpen = childrenOrders.some((order) => order.status === OrderStatus.PENDING)
+  const hasCancelling = childrenOrders.some((order) => order.isCancelling)
+  const hasFilled = childrenOrders.some((order) => order.status === OrderStatus.FULFILLED)
+  const hasCancelled = childrenOrders.some((order) => order.status === OrderStatus.CANCELLED)
+  const hasExpired = childrenOrders.some((order) => order.status === OrderStatus.EXPIRED)
+
+  // All parts filled
+  if (childrenOrders.every((order) => order.status === OrderStatus.FULFILLED)) {
+    return TwapOrderStatus.Fulfilled
+  }
+
+  // All parts cancelled
+  if (childrenOrders.every((order) => order.status === OrderStatus.CANCELLED)) {
+    return TwapOrderStatus.Cancelled
+  }
+
+  // All parts expired
+  if (childrenOrders.every((order) => order.status === OrderStatus.EXPIRED)) {
+    return TwapOrderStatus.Expired
+  }
+
+  // At least one part is open
+  if (hasOpen) {
+    return TwapOrderStatus.Pending
+  }
+
+  // One part is cancelling
+  if (hasCancelling) {
+    return TwapOrderStatus.Cancelling
+  }
+
+  // Some filled + some expired/cancelled
+  if (hasFilled && (hasExpired || hasCancelled)) {
+    return TwapOrderStatus.Fulfilled // Partially filled is considered Fulfilled for display
+  }
+
+  // Mixed cancelled & expired only (no fills)
+  if (hasCancelled && hasExpired && !hasFilled) {
```

Comment:
why the hasExpired?

If we cancel and is not expired, we need to see Cancelled, no?